packages/wrangler/src/__tests__/deploy.test.ts

```diff
@@ -4422,6 +4407,231 @@ addEventListener('fetch', event => {});`
+				{ filePath: "boop/file-2.txt", content: "Content of file-2" },
+			];
+			writeAssets(assets);
+			writeWranglerToml({
+				experimental_assets: { directory: "assets" },
+			});
+			writeWorkerSource();
+			msw.use(
+				http.post(
+					`*/accounts/some-account-id/workers/scripts/test-name/assets-upload-session`,
+					async ({ request }) => {
+						expect(await request.json()).toStrictEqual({
+							manifest: {
+								"file-1.txt": {
+									hash: "0de3dd5df907418e9730fd2bd747bd5e",
+									size: 17,
+								},
+								"boop/file-2.txt": {
+									hash: "7574a8cd3094a050388ac9663af1c1d6",
+									size: 17,
+								},
+							},
+						});
+						return HttpResponse.json(
+							{
+								success: true,
+								errors: [],
+								messages: [],
+								result: { jwt: "<<aus-token>>", buckets: [] },
+							},
+							{ status: 201 }
+						);
+					}
+				)
+			);
+			// skips asset uploading since empty buckets returned
+			mockSubDomainRequest();
+			mockUploadWorkerRequest({
+				expectedMainModule: "no-op-worker.js",
+			});
+			await runWrangler("deploy");
+		});
+
+		it("should upload assets in the requested buckets", async () => {
+			const assets = [
+				{ filePath: "file-1.txt", content: "Content of file-1" },
+				{ filePath: "boop/file-2.txt", content: "Content of file-2" },
+				{ filePath: "boop/file-3.txt", content: "Content of file-3" },
+				{ filePath: "file-4.txt", content: "Content of file-4" },
+				{ filePath: "beep/file-5.txt", content: "Content of file-5" },
+				{ filePath: "beep/file-6.txt", content: "Content of file-1" },
+			];
+			writeAssets(assets);
+			writeWranglerToml({
+				experimental_assets: { directory: "assets" },
+			});
+			writeWorkerSource();
+
+			const uploadRequests: StrictRequest<UploadPayloadFile[]>[] = [];
+			const bodies: UploadPayloadFile[][] = [];
+			const mockBuckets = [
+				[
+					"0de3dd5df907418e9730fd2bd747bd5e",
+					"7574a8cd3094a050388ac9663af1c1d6",
+				],
+				["ff5016e92f039aa743a4ff7abb3180fa"],
+				["f05e28a3d0bdb90d3cf4bdafe592488f"],
+				["0de3dd5df907418e9730fd2bd747bd5e"],
+			];
+			msw.use(
+				http.post(
+					`*/accounts/some-account-id/workers/scripts/test-name/assets-upload-session`,
+					async () => {
+						return HttpResponse.json(
+							{
+								success: true,
+								errors: [],
+								messages: [],
+								result: {
+									jwt: "<<aus-token>>",
+									buckets: mockBuckets,
+								},
+							},
+							{ status: 201 }
+						);
+					}
+				),
+				http.post(
+					"*/accounts/some-account-id/workers/assets/upload",
+					async ({ request }) => {
+						uploadRequests.push(request as StrictRequest<UploadPayloadFile[]>);
+						const body = (await request.text())
+							.split("\n")
+							.map((x) => JSON.parse(x)) as UploadPayloadFile[];
+
+						bodies.push(body);
+						expect(request.headers.get("Authorization")).toBe(
+							"Bearer <<aus-token>>"
+						);
```

Comment:
I know this already exists around the codebase but I don't love this usage of `expect` inside the mock endpoints. When the test fails (it is a network error to the `fetch` call) it makes debugging incredibly difficult.

We should be writing our tests so that they fail in a way that helps us find the cause easily. (Contrived example: using `expect(string).toContain(substring)` vs expect(string.includes(substring)).toBe(true)` results in a better output from vitest that shows both strings and a diff instead of "expected false but got true")

Instead, can we store an array of these Authorization header values in the test function scope and push to it here (like `bodies`) and assert with all the other assertions at the end of the test. (Looks like we already have `uploadedRequests` so use that further down)

Reply:
Ah I agree completely, I meant to fix this yesterday and just forgot 🤦 - will change now